packages/bloc/lib/src/bloc.dart

```diff
@@ -142,39 +154,42 @@ abstract class Bloc<Event, State> extends Stream<State> implements Sink<Event> {
   Stream<State> mapEventToState(Event event);
 
   /// Transforms the `Stream<State>` into a new `Stream<State>`.
-  /// By default [transformStates] returns the incoming `Stream<State>`.
-  /// You can override [transformStates] for advanced usage in order to
+  /// By default [transformTransitions] returns the incoming `Stream<State>`.
+  /// You can override [transformTransitions] for advanced usage in order to
   /// manipulate the frequency and specificity at which `transitions`
   /// (state changes) occur.
   ///
   /// For example, if you want to debounce outgoing [states]:
   ///
   /// ```dart
   /// @override
-  /// Stream<State> transformStates(Stream<State> states) {
-  ///   return states.debounceTime(Duration(seconds: 1));
+  /// Stream<Transition<Event, State>> transformTransitions(
+  ///   Stream<Transition<Event, State>> transitions,
+  /// ) {
+  ///   return transitions.debounceTime(Duration(seconds: 1));
   /// }
   /// ```
-  Stream<State> transformStates(Stream<State> states) => states;
-
-  void _bindStateSubject() {
-    Event currentEvent;
-
-    transformStates(transformEvents(_eventSubject, (event) {
-      currentEvent = event;
-      return mapEventToState(currentEvent).handleError(_handleError);
-    })).forEach(
-      (nextState) {
-        if (state == nextState || _stateSubject.isClosed) return;
-        final transition = Transition(
-          currentState: state,
-          event: currentEvent,
-          nextState: nextState,
-        );
+  Stream<Transition<Event, State>> transformTransitions(
+    Stream<Transition<Event, State>> transitions,
+  ) {
+    return transitions;
+  }
+
+  void _bindEventsToStates() {
+    transformTransitions(
+      transformEvents(_eventController.stream, (event) {
+        return mapEventToState(event)
+            .transform(_TransitionStreamTransformer(this, event))
```

Comment:
Can you just use `map` operator in this case? `_TransitionStreamTransformer` looks complicated.

```
.map((nextState) => 
  Transition(
    currentState: _state,
    event: event,
    nextState: nextState
  )
)
```

Reply:
yup good point. I originally had it due to the `StartWith` changes but it's no longer needed 👍